### ISSUE TITLE
Custom HTTP queries for webhooks

### ISSUE THREAD

**Issue by hooksie1** [gristlabs/grist-core#827]
I think it would be an awesome feature to allow custom headers in webhooks. The main thought is this would let us use systems that require some kind of authentication for the webhooks.

Talked about this in discord, tagging @vviers.

**Comment by vviers** [comment]
we plan on working on this

**Comment by paulfitz** [comment]
Neat! A workaround is passing requests through an external service like cloudflare workers to rewrite them, but that is a pain to setup and then annoying to maintain.

**Comment by CamilleLegeron** [comment]
I made this PR to add headers in webhook settings, I put some screencast in the description
https://github.com/gristlabs/grist-core/pull/832

**Comment by paulfitz** [comment]
With @CamilleLegeron working on webhooks in #832, I wonder if we could sneak a field for this into the migration she'll be introducing. Just a generic `options` column would do, we wouldn't have to commit to any representation yet. We typically have an `options` columns on tables for non-relational stuff that shouldn't require every Grist document everywhere in the world to be migrated when something is added, but there isn't one yet on the `_grist_Triggers` table. Doing it now would save another migration later.

**Comment by CamilleLegeron** [comment]
What about security if there is an authorization secret ? 
Inside webhook settings page maybe add an input for secret and another for none secret ? And that way we can encrypte the secret and set both in options column

**Comment by CamilleLegeron** [comment]
Thinking about what headers could be great to custom, I'm only thinking of the authorization parameter. How about adding only `Authorization` input in the webhooks params and see later if other needs arise?

**Comment by paulfitz** [comment]
> Thinking about what headers could be great to custom, I'm only thinking of the authorization parameter. How about adding only Authorization input in the webhooks params and see later if other needs arise?
Wdyt @paulfitz ?

There's nothing very special to Grist about webhooks. Could be worth looking at what similar apps offer, to see if there is any consensus?

> What about security if there is an authorization secret ?
> Inside webhook settings page maybe add an input for secret and another for none secret ? And that way we can encrypt the secret and set both in options column

The security question is interesting. The webhook URL itself may have a sensitive key in it, which is part of why it is not stored directly in the document but in a `secrets` table in the home database. This makes some things more awkward and some things simpler., and makes adding more encryption in the future easier. It could make sense to treat material for `Authorization` header in the same way, at least partially.

I wonder if this needs actually two features:
* A way to store and retrieve secrets associated with a particular document (the `secrets` table is intended for that purpose)
* A way to include a secret in a header, for example with some kind of syntax that could later extend to pulling information from the payload itself or elsewhere? Like `Authorization: {secrets.AUTH_SECRET}` or something. Maybe this is too ambitious. But it would mean the header configuration itself wouldn't have to be treated as a secret.

**Comment by fflorent** [comment]
I tend to think it be more interesting to change this issue title so it tracks three similar features that will probably be implemented altogether (otherwise please mention your disagreement about that), which are the ability to customize the following things for webhooks:
 - the method;
 - the body;
 - the headers;

([as stated in an above comment](https://github.com/gristlabs/grist-core/issues/827#issuecomment-1895368538))

**Comment by nicolas-imbert** [comment]
Hey, 

Following a quick talk with @fflorent and some users, I just wanted to add my 2 cents on the potential use cases for this feature. 

1. Send an email when there is a new record or updated record (via tools like Brevo or Mailchimp)
2. Send a notification when new record or updated record (via tools like Mattermost or Teams)

And probably a lot more to come !
Looking forward to it, thanks for the job done.

**Comment by fflorent** [comment]
Regarding the implementation, what about allowing configuring a column in which it is expected to return these items thanks to a formula, like:
```python
import json
return {
  'method': 'PATCH', # Defaults to Post
  'headers': { # Headers to append to the one sent by the webhooks (like the headers)
    'X-Specific-Header': $someColumn,
    'X-Common-Header': 'Common value for all records',
  },
  'body': json.dumps({ # the body, defaults to nothing
    'whatever': $someOtherColumn
  }),
  'queryParams': 'blah=' + $queryParams # Defaults to empty string.
  # Other stuffs?
}
```

(The `queryParam` property came to my mind when writing this formula)

What do you think?

**Comment by fflorent** [comment]
We would need to clarify the User Story, prevent security issues and have a design mockup for this issue.

**Comment by vviers** [comment]
Tracking progress here :  
- [x] add Authorisation header in webhooks (#941)
- [ ] allow user to specify the shape of a header sent via webhooks
- [ ] allow user to specify the shape of a payload sent via webhooks
- [ ] allow user to specify query params sent via webhooks
- [ ] allow user to specify the HTTP verb used by webhooks

@hooksie1 could you please rename this issue to the more generic "Custom HTTP queries for webhooks"

**Comment by vviers** [comment]
We tried to make sense of this issue with @hexaltation, @fflorent and @jonathanperret today after spending a lot of time trying to think of what was the simplest way to implement this.

**_TL;DR : after digging into what our users actually want this feature for, we decided our energy would be put to better use by making other ongoing PRs and issues land._**

### What do we actually want ?

We did a little bit of user research to understand what people actually needed to "customize webhooks" for, and basically the answer was :
- the vast majority of them **want some sort of notification** when a new record is added to their tables (most of the time via an external form). This could obviously be done by email or some sort of slack-like channel, but **most people would already be very happy if they could be notified directly in grist**.
- some of them want to **automatically push data from Grist into another tool** (in our context, https://www.demarches-simplifiees.fr/?locale=en comes up a lot for example)

### How do we get there ?

The two above needs could actually be met in a much simpler way through other means : 
- ⏫ making it easier and safer for people to use automation tools such as n8n or ActivePieces. To us, the main hurdle to encouraging more and more of our users to use those tools is that [Service Accounts](https://github.com/gristlabs/grist-core/issues/1217) have not landed yet. Crucially, **we don't think Grist itself should be an automations tool**, and should let the burden of making POST requests to external APIs to other dedicated tools. In Airtable for example, the Automations toolbox feels like a specific project in and of itself (looking more like Zappier than an online spreadsheet tool) and is only loosely coupled to their main database/spreadsheet product.
- ☁ activating the `REQUESTS` function on our instance, which will be safer once https://github.com/gristlabs/grist-core/pull/1363 lands. This will let advanced users who want to tinker with calling external services from their grist tables do it (⚠ although we don't think we should encourage this until we have a clean way to store secrets in a grist document ⚠) 
- 🔔 making the [Notifications project](https://github.com/gristlabs/grist-core/issues/1086) land, which could then be followed up with developping https://github.com/gristlabs/grist-core/issues/146

To summarize we propose to close this issue in favor of the following :
- finishing https://github.com/gristlabs/grist-core/issues/1217
- finishing https://github.com/gristlabs/grist-core/pull/1363 (along with generalizing what was done in https://github.com/gristlabs/grist-core/pull/941 to make secrets in the secrets table available to functions ?)
- starting to work on https://github.com/gristlabs/grist-core/issues/1086, we think we'll split up the notifications project into more digestible chunks (Comments, Form submissions, Document invitations, etc.)

**Comment by vviers** [comment]
Woops, closed this too fast maybe as being able to customize the HEADER of webhooks could still be a valid need, we're just backing-off from implementing the following : 

WONTDO :
allow user to specify the shape of a payload sent via webhooks
allow user to specify query params sent via webhooks
allow user to specify the HTTP verb used by webhooks

We can talk about whether we think "allow user to specify the shape of a header sent via webhooks" should be implemented, but then I thnk opening a dedicated issue with a user story would be more useful than this umbrella issue which got too big and messy (sorry abut this @hooksie1, this was your original need all along 😬 )

**Comment by hooksie1** [comment]
No worries! Yeah this did get bigger than I expected 😂

**Comment by jeffam** [comment]
I came across this issue after searching for a way to determine the document id and table name that triggered the webhook. In my case, my webhook looks at the data in the payload and then updates some or all records based on that.

If the webhook POST request contained the grist document id and triggering table name in, say, the POST request headers, I wouldn't have to manually configure those in my webhook processor.
